sharding/collation.go

```diff
@@ -5,43 +5,97 @@ import (
+	Period            *big.Int        // the period number in which collation to be Pncluded.
+	ProposerAddress   *common.Address // address of the collation proposer.
+	ProposerSignature []byte          // the proposer's signature for calculating collation hash.
+}
 
-// Transactions returns an array of tx's in the collation.
-func (c *Collation) Transactions() []*types.Transaction { return c.transactions }
+// NewCollationHeader initializes a collation header struct.
+func NewCollationHeader(shardID *big.Int, chunkRoot *common.Hash, period *big.Int, proposerAddress *common.Address, proposerSignature []byte) *CollationHeader {
+	data := collationHeaderData{
+		ShardID:           shardID,
+		ChunkRoot:         chunkRoot,
+		Period:            period,
+		ProposerAddress:   proposerAddress,
+		ProposerSignature: proposerSignature,
+	}
+	return &CollationHeader{data}
+}
+
+// Hash takes the keccak256 of the collation header's contents.
+func (h *CollationHeader) Hash() (hash common.Hash) {
+	hw := sha3.NewKeccak256()
+	rlp.Encode(hw, h)
+	hw.Sum(hash[:0])
+	return hash
+}
 
 // ShardID is the identifier for a shard.
-func (c *Collation) ShardID() *big.Int { return c.header.shardID }
+func (h *CollationHeader) ShardID() *big.Int { return h.data.ShardID }
 
 // Period the collation corresponds to.
-func (c *Collation) Period() *big.Int { return c.header.period }
+func (h *CollationHeader) Period() *big.Int { return h.data.Period }
 
-// ProposerAddress is the coinbase addr of the creator for the collation.
-func (c *Collation) ProposerAddress() *common.Address { return c.header.proposerAddress }
+// ChunkRoot of the serialized collation body.
+func (h *CollationHeader) ChunkRoot() *common.Hash { return h.data.ChunkRoot }
+
+// EncodeRLP gives an encoded representation of the collation header.
+func (h *CollationHeader) EncodeRLP() ([]byte, error) {
+	encoded, err := rlp.EncodeToBytes(&h.data)
+	if err != nil {
+		return nil, err
+	}
+	return encoded, nil
+}
+
+// DecodeRLP uses an RLP Stream to populate the data field of a collation header.
+func (h *CollationHeader) DecodeRLP(s *rlp.Stream) error {
+	err := s.Decode(&h.data)
+	return err
+}
+
+// Header returns the collation's header.
+func (c *Collation) Header() *CollationHeader { return c.header }
+
+// Body returns the collation's byte body.
+func (c *Collation) Body() []byte { return c.body }
+
+// Transactions returns an array of tx's in the collation.
+func (c *Collation) Transactions() []*types.Transaction { return c.transactions }
 
-// SetHeader updates the collation's header.
-func (c *Collation) SetHeader(h *CollationHeader) { c.header = h }
+// ProposerAddress is the coinbase addr of the creator for the collation.
+func (c *Collation) ProposerAddress() *common.Address {
+	return c.header.data.ProposerAddress
+}
 
 // AddTransaction adds to the collation's body of tx blobs.
 func (c *Collation) AddTransaction(tx *types.Transaction) {
 	// TODO: Include blob serialization instead.
```

Comment:
github issue here too

Reply:
Use blob serialization instead of appending tx to c.transactions `c.transactions = append(c.transactions, tx)`